File: python/rlox/reward_shaping.py
"""Potential-based reward shaping (PBRS) backed by Rust.

Provides composable reward shaping that preserves optimal policy invariance
(Ng et al., 1999). The heavy arithmetic is done in Rust via
``rlox.shape_rewards_pbrs`` and ``rlox.compute_goal_distance_potentials``.
"""

from __future__ import annotations

from collections.abc import Callable

import numpy as np

import rlox._rlox_core as _core


class PotentialShaping:
    """PBRS reward shaping with a configurable potential function.

    Computes: r' = r + gamma * Phi(s') * (1 - done) - Phi(s)

    The potential function can be a neural network, a distance function,
    or any callable ``(observations) -> potentials``.

    Parameters
    ----------
    potential_fn : callable
        Maps (N, obs_dim) ndarray -> (N,) ndarray of potential values.
    gamma : float
        Discount factor (default 0.99).
    """

    def __init__(self, potential_fn: Callable[[np.ndarray], np.ndarray], gamma: float = 0.99) -> None:
        self.potential_fn = potential_fn
        self.gamma = gamma

    def shape(
        self,
        rewards: np.ndarray,
        obs: np.ndarray,
        next_obs: np.ndarray,
        dones: np.ndarray,
    ) -> np.ndarray:
        """Compute PBRS shaped rewards.

        Parameters
        ----------
        rewards : (N,) base rewards
        obs : (N, obs_dim) current observations
        next_obs : (N, obs_dim) next observations
        dones : (N,) episode termination flags (1.0 = done)

        Returns
        -------
        shaped_rewards : (N,) float64 array
        """
        phi_current = np.ascontiguousarray(self.potential_fn(obs), dtype=np.float64)
        phi_next = np.ascontiguousarray(self.potential_fn(next_obs), dtype=np.float64)
        rewards = np.ascontiguousarray(rewards, dtype=np.float64)
        dones = np.ascontiguousarray(dones, dtype=np.float64)

        return _core.shape_rewards_pbrs(rewards, phi_current, phi_next, self.gamma, dones)


class GoalDistanceShaping:
    """Goal-distance potential shaping backed by Rust.

    Potential: Phi(s) = -scale * ||s[goal_slice] - goal||_2

    Moving closer to the goal increases the potential, producing a
    positive shaping bonus. Moving farther away decreases it.

    Parameters
    ----------
    goal : (goal_dim,) target goal vector
    obs_dim : total observation dimensionality
    goal_start : index where goal-relevant dims start in obs
    goal_dim : number of goal-relevant dimensions
    scale : scaling factor for the distance potential
    gamma : discount factor (default 0.99)
    """

    def __init__(
        self,
        goal: np.ndarray,
        obs_dim: int,
        goal_start: int,
        goal_dim: int,
        scale: float = 1.0,
        gamma: float = 0.99,
    ) -> None:
        self.goal = np.asarray(goal, dtype=np.float64)
        self.obs_dim = obs_dim
        self.goal_start = goal_start
        self.goal_dim = goal_dim
        self.scale = scale
        self.gamma = gamma

    def shape(
        self,
        rewards: np.ndarray,
        obs: np.ndarray,
        next_obs: np.ndarray,
        dones: np.ndarray,
    ) -> np.ndarray:
        """Compute goal-distance shaped rewards.

        Parameters
        ----------
        rewards : (N,) base rewards
        obs : (N, obs_dim) current observations
        next_obs : (N, obs_dim) next observations
        dones : (N,) episode termination flags

        Returns
        -------
        shaped_rewards : (N,) float64 array
        """
        obs_2d = np.asarray(obs, dtype=np.float64)
        next_obs_2d = np.asarray(next_obs, dtype=np.float64)
        rewards = np.ascontiguousarray(rewards, dtype=np.float64)
        dones = np.ascontiguousarray(dones, dtype=np.float64)

        # Rust expects flat (N * obs_dim,) arrays
        obs_flat = np.ascontiguousarray(obs_2d.ravel(), dtype=np.float64)
        next_obs_flat = np.ascontiguousarray(next_obs_2d.ravel(), dtype=np.float64)

        phi_current = _core.compute_goal_distance_potentials(
            obs_flat, self.goal, self.obs_dim, self.goal_start, self.goal_dim, self.scale,
        )
        phi_next = _core.compute_goal_distance_potentials(
            next_obs_flat, self.goal, self.obs_dim, self.goal_start, self.goal_dim, self.scale,
        )

        phi_current = np.ascontiguousarray(phi_current)
        phi_next = np.ascontiguousarray(phi_next)

        return _core.shape_rewards_pbrs(rewards, phi_current, phi_next, self.gamma, dones)

File: examples/rust/src/bin/reward_shaping.rs
//! Potential-based reward shaping (PBRS) demonstration.
//!
//! Shows how shaped rewards preserve the optimal policy
//! while accelerating learning.
//!
//! ```bash
//! cd examples/rust && cargo run --bin reward_shaping
//! ```

use rlox_core::training::reward_shaping::{shape_rewards_pbrs, compute_goal_distance_potentials};

fn main() {
    // --- PBRS with custom potentials ---
    let rewards = vec![1.0, 1.0, 1.0, 1.0, 0.0];
    let phi_current = vec![0.5, 0.6, 0.7, 0.8, 0.9]; // potential of current state
    let phi_next = vec![0.6, 0.7, 0.8, 0.9, 0.0];     // potential of next state
    let gamma = 0.99;
    let dones = vec![0.0, 0.0, 0.0, 0.0, 1.0];

    let shaped = shape_rewards_pbrs(&rewards, &phi_current, &phi_next, gamma, &dones).unwrap();

    println!("PBRS: r' = r + γΦ(s') - Φ(s)");
    println!("Step | Raw    | Shaped | Φ(s)  | Φ(s')");
    println!("-----|--------|--------|-------|------");
    for i in 0..rewards.len() {
        println!(
            "  {i}  | {:.3}  | {:.3}  | {:.2}  | {:.2}{}",
            rewards[i],
            shaped[i],
            phi_current[i],
            phi_next[i],
            if dones[i] == 1.0 { "  (done — raw reward only)" } else { "" }
        );
    }

    // --- Goal-distance potentials ---
    println!("\nGoal-distance potentials: Φ(s) = -scale * ||s - goal||");

    let obs_dim = 3;
    let observations = vec![
        0.0, 0.0, 0.0, // far from goal
        0.5, 0.5, 0.5, // closer
        0.9, 0.9, 0.9, // very close
        1.0, 1.0, 1.0, // at goal
    ];
    let goal = vec![1.0, 1.0, 1.0];

    let potentials = compute_goal_distance_potentials(
        &observations,
        &goal,
        obs_dim,
        0,         // goal_start: goal-relevant dims start at index 0
        obs_dim,   // goal_dim: all 3 dims are goal-relevant
        1.0,       // scale
    ).unwrap();

    println!("Obs         | Distance | Potential");
    println!("------------|----------|----------");
    for i in 0..4 {
        let obs = &observations[i * obs_dim..(i + 1) * obs_dim];
        let dist: f64 = obs.iter().zip(goal.iter()).map(|(o, g)| (o - g).powi(2)).sum::<f64>().sqrt();
        println!("  [{:.1}, {:.1}, {:.1}] | {dist:.3}    | {:.3}", obs[0], obs[1], obs[2], potentials[i]);
    }
    println!("\nCloser to goal → less negative potential → higher shaped reward");
}
